J-PAKE interface definition for PSA Crypto

yanesca opened the pull request:
Supporting documentation/design considerations:

- [pake_protocols_to_consider_v1.1.pdf](https://github.com/ARMmbed/mbedtls/files/6279167/pake_protocols_to_consider_v1.1.pdf)
- [pake_abstract_interface_v1.1.pdf](https://github.com/ARMmbed/mbedtls/files/6279171/pake_abstract_interface_v1.1.pdf)
- [pake_interface_design_v1.1.pdf](https://github.com/ARMmbed/mbedtls/files/6279172/pake_interface_design_v1.1.pdf)

Fixes #4000 .

stevew817 commented on include/psa/crypto_types.h:
```diff
@@ -380,4 +380,54 @@ typedef uint16_t psa_key_derivation_step_t;
+ * Encodes which side of the protocol is being executed. For more information
+ * see the documentation of individual PSA_PAKE_SIDE_XXX constants.
+ */
+typedef uint16_t psa_pake_side_t;
```
Is there a reason for making this a 16-bit instead of an 8-bit type? I see the `PSA_PAKE_SIDE_XXX` constants really only use 2 bits of information at the moment: whether the algorithm is augmented, and whether the 'side' is client/first or server/second

yanesca replied:
Good point, `uint8_t` should be enough.

stevew817 commented on include/psa/crypto_values.h:
```diff
@@ -1678,6 +1691,58 @@
+ * creating the cipher suite.
+ *
+ * In theory the protocol works with any non-interactive zero-knowledge proof.
+ * Implementations of the present specification use Schnorr NIZK and this does
```
```suggestion
 * Implementations of the present specification use Schnorr NIZKP and this does
```

stevew817 commented on include/psa/crypto_values.h:
```diff
@@ -2011,4 +2076,175 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+
+/** The PAKE uses elliptic curves.
+ *
+ * The corresponding family type is ::psa_ecc_family_t. in determining a
```
```suggestion
 * The corresponding family type is ::psa_ecc_family_t. In determining a
```

stevew817 commented on include/psa/crypto_values.h:
```diff
@@ -2011,4 +2076,175 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ *                      on \p type, for more information consult the
+ *                      documentation of individual ::psa_pake_primitive_type_t
+ *                      constants).
+ * \param bits          The bitwise of the primitive
```
```suggestion
 * \param bits          Value of type ::psa_pake_bits_t. The interpretation
```

yanesca replied:
This was indeed a typo, but I think that it is better to fix it than to delete the whole line.

stevew817 commented on include/psa/crypto_values.h:
```diff
@@ -2011,4 +2076,175 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ *                      documentation of individual ::psa_pake_primitive_type_t
+ *                      constants).
+ * \param bits          The bitwise of the primitive
+ *                      (Value of type ::psa_pake_bits_t. The interpretation
```
```suggestion

```

stevew817 commented on include/psa/crypto_values.h:
```diff
@@ -2011,4 +2076,175 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ *                      (Value of type ::psa_pake_bits_t. The interpretation
+ *                      of this parameter depends on \p family, for more
+ *                      information consult the documentation of individual
+ *                      ::psa_pake_primitive_type_t constants).
```
```suggestion
 *                      ::psa_pake_primitive_type_t constants.
```

stevew817 commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3759,418 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *
+ * Before calling any function on a PAKE operation object, the application
+ * must initialize it by any of the following means:
+ * - Set the structure to all-bits-zero, for example:
```
Isn't this implicit for all operation structures used by PSA multi-part APIs? If it doesn't need to be explicitly defined for other structures, then what is so different about PAKE? Alternatively, if it does need explicit definition, shouldn't that be added for all other structures as well?

yanesca replied:
I think this is explicitly defined for all other structures, I added this to be consistent with the others. As far as I can tell, hash, MAC, cipher, AEAD, and key derivation all document this explicitly.

athoelke replied:
This is common across all operation structures already defined

stevew817 commented on include/psa/crypto_types.h:
```diff
@@ -380,4 +380,54 @@ typedef uint16_t psa_key_derivation_step_t;
+ * The type and family is not enough to identify the primitive to use in the
+ * PAKE, the implementation needs to know the bitsize too.
+ */
+typedef uint16_t psa_pake_bits_t;
```
Since it is declared as being a mirror of `psa_key_bits_t`, why not typedef it to that? Or even just use `psa_key_bits_t` straight-up?

yanesca replied:
I feel uncomfortable about using `psa_key_bits_t` directly, as I mentioned below I think there is value in separating these types, because as far as the PSA Crypto API is concerned, this information is not related to keys:
https://github.com/ARMmbed/mbedtls/pull/4317#discussion_r614098441

Using typedef ties these two types together, which seems to make sense since in the case of primitives involved with J-PAKE the semantics of these two types is tied together. However this is not true in general and later the types might diverge.

Also, the macros handling them are different and if we use typedef, changing `psa_key_bits_t` might cause bugs in the macros using `psa_pake_bits_t` that would be an increased chance of being missed during review.

athoelke replied:
`psa_key_bits_t` is not a type defined by PSA Crypto (even if it does exist in mbedtls). The PSA Cyprto API uses `size_t` for key bit sizes.

If the use of this in pake is to identify a specific family member, then it _is_ a mirror for how the bit size is used in an asymmetric key. For asymmetric keys, the bit-size does not identify the literal size of the key, it specifies the specific member of the key family that is defined by the key type.

stevew817 commented on include/psa/crypto_types.h:
```diff
@@ -380,4 +380,54 @@ typedef uint16_t psa_key_derivation_step_t;
+ * (Options like for example EnvelopeMode in OPAQUE or "Per-User M and N" in
+ * SPAKE2.)
+ */
+typedef uint32_t psa_pake_cipher_suite_options_t;
```
Is this intended to be a bitmask?

yanesca replied:
Yes, I thought of this as being a bitmask, but now that you ask, it doesn't have to be. This could potentially be sized down to `uint8_t`.

stevew817 commented on include/psa/crypto_values.h:
```diff
@@ -2011,4 +2076,175 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ * \return The constructed primitive value.
+ */
+#define PSA_PAKE_PRIMITIVE(type, family, bits) \
+    ((psa_pake_primitive_t) (((type) << 24 | (persistence) << 16) | (bits)))
```
```suggestion
    ((psa_pake_primitive_t) (((type) << 24 | (family) << 16) | (bits)))
```

stevew817 reviewed (commented):
Some minor things, but overall this looks suitable for the purpose to me.

gilles-peskine-arm reviewed (changes requested):
I've done a first round of review. I've reached the point where there's not much more I can say without either diving much more into how JPAKE and other PAKE protocols work or implementing this API.

yanesca commented on include/psa/crypto_values.h:
```diff
@@ -2011,4 +2076,175 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ */
+#define PSA_PAKE_PRIMITIVE_TYPE_FIELD_DH       ((psa_pake_primitive_type_t)0x02)
+
+/** Construct a PAKE primitive from type, family and bitsize.
```
> Why then is it constructed differently?

I am sorry I am not sure I understand, differently from what? Differently from key type? I thought key type does not encode key length.

gilles-peskine-arm replied:
I'm wondering why there is a separate type `psa_pake_primitive_t`, rather than passing a `psa_key_type_t key_agreement_key_type` and a `size_t bits`.

yanesca replied:
Oh I see, one of the reasons is what you guessed, the groups used for key agreement don't cover everything.

For example RSP uses rings instead of groups. The modulus might be similar and technically we might be able to use key agreement key types to represent them, but feels like a hack and might lay the bed for very subtle mistakes.

Even if we look at J-PAKE, we have subtle issues. The key shares might be technically Diffie-Hellman public keys, the ZKP public keys might look like Diffie-Hellman keys and the security of the scheme relies on the Diffie-Hellman assumption (without looking it up I am not sure which one, but my memory is that for key agreement computational is enough, but for the ZKP you need decisional). Even if the security assumptions are aligned, these "keys" are in a complex interaction with each other and I feel uncomfortable using the same type for them as for key agreement.

Another reason is that as far as the interface is concerned this is not a key type. There are no keys of this type at all. As mentioned above, in theory, some of the data exchanged might be considered of this key type, but this is not shown in the interface, because these are not secret values and differentiating data that can be thought of this way in general does come with a cost and doesn't have any benefit.

In summary, technically it would be possible (would potentially introduce new families for key agreement that only would be used in PAKEs), but I think that it is cleaner to separate them.

yanesca commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3759,418 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ * function several times or you might not need to call this at all.
+ *
+ * Calling this function with PSA_PAKE_DATA_KEY_SHARE as \p type is equivalent
+ * to calling psa_pake_set_key_share().
```
I was slightly conflicted by that too and was thinking about removing it. I kept it because:
1. That was the direction we agreed on the latest meeting before I raised this PR
2. It might not make much sense for J-PAKE, but feels appropriate to more mainstream PAKEs

In summary, I am happy to remove it for now, we still can add it back later if we think it is worth it.

yanesca commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3759,418 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *                              value derived from the password (eg. by a
+ *                              memory-hard function).  It must remain valid
+ *                              until the operation terminates. It must be of
+ *                              type PSA_KEY_TYPE_PASSWORD or
```
I removed the reference to make the CI pass, it can be added back once the PBKDF API (which introduces this key type) is merged.

yanesca commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3759,418 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+
+/** Get the key share from a password-authenticated key exchange operation.
+ *
+ * This function returns a simple key share (eg. group element).
```
Key shares are very similar to key shares in key agreement protocols. They are public values that we need to send to the peer. I don't think it makes sense to keep them inside the crypto service boundary.

yanesca commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3759,418 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *                              #psa_pake_operation_t and not yet in use.
+ * \param password              Identifier of the key holding the password or a
+ *                              value derived from the password (eg. by a
+ *                              memory-hard function).  It must remain valid
```
> When would one use a password vs a derived value?

It would be a password if the password is used directly as it is and it would be derived when it went through a memory hard function first.

> Does it depend on the PAKE protocol?

The memory hard function applied to the password (if any) is out of scope for most standards. Some of them mention the possibility, some of them recommend, but are agnostic. This step can be done prior the execution of the PAKE.

The exception is OPAQUE where the memory hard function is integral part of the algorithm (can't be split off as a pre-processing step) and therefore (although technically possible) it doesn't make any sense to use an MHF before passing the password.

yanesca commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3759,418 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *                              type PSA_KEY_TYPE_PASSWORD or
+ *                              #PSA_KEY_TYPE_DERIVE. It has to allow the usage
+ *                              #PSA_KEY_USAGE_DERIVE.
+ * \param alg                   The PAKE protocol to use
```
I am happy to call them "PAKE algorithm". I only called them protocol in the design documents, to emphasise their complexity.

yanesca commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3759,418 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ * \param[in,out] operation     The operation object to set up. It must have
+ *                              been initialized as per the documentation for
+ *                              #psa_pake_operation_t and not yet in use.
+ * \param password              Identifier of the key holding the password or a
```
Both sounds like a good idea, would make the API more convenient to use.

yanesca replied:
Thinking about it a bit more, using the password directly is less secure than pressing it through an MHF first. If it is less secure and less convenient, then it is exactly how we want it and it might be better not make the direct input variant easier.

(Similarly we might want to make it more difficult to store a password directly as a key than something generated by a PBKDF, but this is out of scope for this PR.)

yanesca commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3759,418 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *                              side of the protocol that is being set up. For
+ *                              more information see the documentation of \c
+ *                              PSA_PAKE_SIDE_XXX constants.
+ * \param[in] user_id           The user ID to authenticate with.
```
The length of user ids and the session data can vary across applications independently from the PAKE algorithm used. For most algorithm these are included in the handshake transcript in some way. I wanted to make this a single call, because this way even malloc-free implementations can include rolling hashes in the context instead of the raw ids and session data, which makes the context smaller and eliminates possible limitations on the size of input.

In theory we could achieve the same with individual function calls, but then these would need to be called in a specific order, which might not be the same for different PAKEs.

In summary I am inclined to keeping this a single function call. On the other hand you are making a very good point about providing a way to input not just keys, but raw passwords and key derivation objects.

Adding these would mean having something like this:
```
psa_pake_setup_with_key(&operation, key_id, ...);
psa_pake_setup_with_password(&operation, password, password_len, ...);
psa_pake_setup_with_mhf(&pake_operation, &kdf_operation, password_len, ...);
```

What do you think?

yanesca commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3759,418 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+                            size_t key_share_size,
+                            size_t *key_share_length);
+
+/** Get additional key share from a password-authenticated key exchange.
```
As mentioned bellow, I am happy to remove it for now. The main purpose of `psa_pake_get_key_share()` is to provide a more straightforward and convenient call flow for regular PAKEs. Most PAKEs have a single key share and need a flow like this:

```
op=PSA_PAKE_OPERATION_INIT;
psa_pake_setup();
psa_pake_get_key_share();
psa_pake_set_key_share();
psa_pake_get_implicit_key(); 
```

Adding `psa_pake_get/set_key_share()` functions cuts out the `psa_pake_data_t` constants from the users vision, hiding complexity that exists only for unrelated PAKEs that aren't relevant for the user.

Depending on how you look at it, J-PAKE has somewhere between 3-9 key shares. A rough estimate for OPAQUE is 2-6. The other 4 I studied have only 1. 

The key shares are public values we need to send to the peer and we won't need to add bytes/key/kdf variants.

yanesca commented on include/psa/crypto_values.h:
```diff
@@ -2011,4 +2076,175 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ *                      on \p type, for more information consult the
+ *                      documentation of individual ::psa_pake_primitive_type_t
+ *                      constants).
+ * \param bits          The bitwise of the primitive
```
It is a typo, should be bitsize.

yanesca commented on include/psa/crypto_values.h:
```diff
@@ -1678,6 +1691,58 @@
+ * All the remaining parameters passed to psa_pake_cipher_suite() when creating
+ * the cipher suite must be 0.
+ *
+ * The key exchange flow for JPAKE is as follows:
```
Sure, good idea, will do.

yanesca commented on include/psa/crypto_types.h:
```diff
@@ -380,4 +380,54 @@ typedef uint16_t psa_key_derivation_step_t;
+/** Encoding of the bitsize for the PAKE's primitive.
+ *
+ * The type and family is not enough to identify the primitive to use in the
+ * PAKE, the implementation needs to know the bitsize too.
```
Good point, I wasn't thinking that far ahead.

yanesca replied:
How this is best implemented depends on the outcome of these two other discussions:
https://github.com/ARMmbed/mbedtls/pull/4317#discussion_r613511051
https://github.com/ARMmbed/mbedtls/pull/4317#discussion_r613473125

I'll wait with addressing this until  the above two conversations resolve.

yanesca replied:
This type has been removed and the cipher suite now takes a `size_t` for the bit size.

yanesca commented on include/psa/crypto_values.h:
```diff
@@ -1678,6 +1691,58 @@
+ * For more information consult the documentation of the individual
+ * PSA_PAKE_DATA_XXX constants.
+ *
+ * J-PAKE is standardised for example in RFC 8236 and in THREAD.
```
Sure, will do.

yanesca replied:
The THREAD specification seems to be restricted to the members of the THREAD group and I can't give a similarly obvious pointer here. I don't see any compelling reason to mention a standard that is not publicly available, I will remove the reference to THREAD.

yanesca commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3759,418 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *
+ * This is an implementation-defined \c struct. Applications should not
+ * make any assumptions about the content of this structure except
+ * as directed by the documentation of a specific implementation.
```
This wording is still present in at least 6 other documentation sections. I am happy to change it, but would prefer to agree on the exact wording before I do.

Are you happy with this wording?
> This is an implementation-defined \c struct. Applications should not
    make any assumptions about the content of this structure.
    Implementation details can change in future versions without notice.

yanesca commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3759,418 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *
+ * \retval The constructed cipher suite.
+ */
+static psa_pake_cipher_suite_t psa_pake_cipher_suite(
```
EC J-PAKE is quite flexible regarding cipher suites, but many PAKEs only have a handful of them. For these it would be convenient to have the ability to define constants for cipher suites that can be passed to the setup function directly.

For example:

```c
#define SPAKE2_P256_SHA256_HMAC                       \
psa_pake_cipher_suite(                                \
    PSA_PAKE_SPAKE2,                                  \
    PSA_PAKE_PRIMITIVE(                               \
        PSA_PAKE_PRIMITIVE_TYPE_CURVE,                \
        PSA_ECC_FAMILY_SECP_R1,                       \
        256),                                         \
    PSA_ALG_SHA_256,                                  \
    PSA_ALG_HMAC_BASE,                                \
    0, 0, 0, 0)
```
(In this example I already incorporated your suggestion below - https://github.com/ARMmbed/mbedtls/pull/4317#discussion_r613476966)

This could be useful for J-PAKE as well since protocols using it have to agree on it a-priori or negotiate it. Either way there is a limited set of cipher suites and it would make sense to define them as such constants.

yanesca replied:
I have reworked this part as we agreed to use getters and setters. (We can add convenience macros later.) I also only added setter/getter functions for the attributes needed by JPAKE. Further functions can be added when needed as the structure is implementation defined.

yanesca commented:
Hi @stevew817 and @gilles-peskine-arm, thank you very much for your reviews!

I believe I addressed everything I could without receiving further feedback. Could you please have a look at the new commits and progress the discussions we started?

mpg commented on tmp:
```diff
@@ -0,0 +1,6 @@
+k=n-mt
```
Was it intentional to commit this file?

yanesca replied:
No, that was by accident. Thanks for catching.

athoelke commented on include/psa/crypto_values.h:
```diff
@@ -1678,6 +1691,78 @@
+ *
+ * J-PAKE is standardised for example in RFC 8236.
+ */
+#define PSA_ALG_PAKE_JPAKE                   ((psa_algorithm_t)0x0a000001)
```
In the algorithm encoding used by PSA Crypto, bits[7:0] are _always_ used to identify a hash algorithm.

I suspect that the `0x01` value used in this encoding is to identify the type of PAKE algorithm? In which case, unless there is a reason to use bits[15:8] for something else, it would be more consistent to use bits[15:8] for this information, and leave bits[7:0] as zero.

athoelke replied:
Is the use of a secure hash algorithm common within PAKE constructions? - If so, is it worth encoding the hash used with J-PAKE into the algorithm identifier?

An alternative approach to the use of the algorithm identifier encoding space would be to do something similar to the key-agreement + KDF, and allow a composite PAKE + KDF algorithm identifier (as well as the raw PAKE)?

yanesca replied:
I have updated the value. I didn't encode the hash in the last byte, I don't think it would worth it, because it is never enough to fully characterise a PAKE.

athoelke commented on include/psa/crypto_values.h:
```diff
@@ -2011,4 +2096,176 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ *  curve would be. For more information, consult the documentation of
+ *  psa_export_key().
+ */
+#define PSA_PAKE_PRIMITIVE_TYPE_CURVE       ((psa_pake_primitive_type_t)0x01)
```
In the key types API, PSA Crypto uses `ECC` to denote Elliptic curve cryptography.

athoelke commented on include/psa/crypto_values.h:
```diff
@@ -2011,4 +2096,176 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ *  group would be. For more information, consult the documentation of
+ *  psa_export_key().
+ */
+#define PSA_PAKE_PRIMITIVE_TYPE_FIELD_DH       ((psa_pake_primitive_type_t)0x02)
```
In the key types API, PSA Crypto uses `DH` to denote Finite Field Diffie-Hellman.

athoelke commented on include/psa/crypto_values.h:
```diff
@@ -2011,4 +2096,176 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ */
+#define PSA_PAKE_PRIMITIVE(pake_type, pake_family, pake_bits) \
+    ((psa_pake_primitive_t) (((pake_type) << 24 |             \
+            (pake_family) << 16) | (pake_bits)))
```
This packing is what drives the 16-bit `psa_pake_bits_t` definition. I suspect that the encoding of the pake primitive could be implementation specific - and does not need to be standardised by PSA Crypto?

yanesca replied:
Yes, it was and it wasn't intended to be standardised by PSA. Also, I reworked it as we agreed and now these types are not present anymore.

yanesca commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3759,418 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *
+ * \retval The constructed cipher suite.
+ */
+static psa_pake_cipher_suite_t psa_pake_cipher_suite(
```
Thinking about it a bit more I didn't see any value in either the redundancy or aligning to the rest of the API in passing a naked algorithm identifier and in the new version went for option 3.

yanesca commented:
I have updated the PR based on what we discussed. There still two ongoing discussions that need to be resolved before moving forward:
https://github.com/ARMmbed/mbedtls/pull/4317#discussion_r613511051
https://github.com/ARMmbed/mbedtls/pull/4317#discussion_r613511537

gilles-peskine-arm commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3760,483 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ * state and must be aborted by calling psa_pake_abort().
+ *
+ * \param[in,out] operation    Active PAKE operation.
+ * \param[out] output          A key derivation operation that has been
+ *                             initialized and set up.
```
```suggestion
 * \param[out] output          A key derivation operation that is ready
 *                             for an input step of type
 *                             #PSA_KEY_DERIVATION_INPUT_SECRET.
```

gilles-peskine-arm commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3760,483 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ * \param[in,out] operation    Active PAKE operation.
+ * \param type                 The type of the data that is requested.
+ * \param[out] output          Buffer where the output is to be written.
+ * \param output_size          Size of the \p output buffer in bytes.
```
This should mention `PSA_PAKE_OUTPUT_SIZE`.

gilles-peskine-arm commented on include/psa/crypto_sizes.h:
```diff
@@ -1086,4 +1086,33 @@
+ *                      the output type or PAKE algorithm is not recognized, or
+ *                      the parameters are incompatible, return 0.
+ */
+#define PSA_PAKE_OUTPUT_SIZE(alg, cipher_suite, output)
```
I hadn't thought of this, but now that `cipher_suite` is a structure, it can't easily be a compile-time constant anymore. Can we still have compile-time-constant cipher suite structures such that a call to this macro can evaluate to a compile-time constant?

athoelke replied:
Structures can't be compile-time constants - so the above macro cannot generate a compile-time constant used to dimension an array at compile time.

The cipher-suite needs to be encoded into a scalar (ideally integral) value/values to be usable as a compile-time constant.

One approach would be to not configure the output size on the cipher suite, but just provide a sufficient size based on the algorithm alone. This could be 'close enough' to avoid the complexity of runtime allocation.

An extension of that would be to have `PSA_PAKE_OUTPUT_SIZE()` take the most significant elements (as regards impact on output size) of the cipher_suite as parameters, and provide macro names to directly provide those elements for the common/standard cipher_suites.

Another approach would be to have the common PAKE cipher suites given a 'PAKE ID' value and use that as a shortcut for applications that want to use a standard construction - which would enable compile-time construction of right-sized output buffers (rather than using a MAX_SIZE buffer). Applications using the built-it-yourself cipher suite structure would have to either use dynamic allocation (via a non-constant `psa_pake_output_size()` function) or go for a MAX_SIZE buffer.

yanesca replied:
How about going back to having a psa_pake_primitive_t member replacing the type, family and bits trio in the ciphersuite?

Then this macro could take a primitive and still provide a useful size in compile time.

From the PSA Spec point of view:
- this type would be opaque, but with the restriction that implementations must realise it as an integral type.
- the spec would still need to define how the user can set a specific primitive, but the constructor macro would translate between the standardised representation and the possibly packed internal representation

What do you think?

gilles-peskine-arm commented on include/psa/crypto_sizes.h:
```diff
@@ -1086,4 +1086,33 @@
+ * \param alg           A PAKE algorithm (PSA_ALG_XXX value such that
+ *                      #PSA_ALG_IS_PAKE(\p alg) is true).
+ * \param cipher_suite  A cipher suite that is compatible with algorithm \p alg.
+ * \param output        An output type used with algorithm \p alg.
```
`output` is usually an output buffer. What's an “output type”? A `psa_pake_data_t`?

gilles-peskine-arm commented on include/psa/crypto_types.h:
```diff
@@ -380,4 +380,31 @@ typedef uint16_t psa_key_derivation_step_t;
+ * This type is for encoding additional input and output data for such
+ * algorithms.
+ */
+typedef uint8_t psa_pake_data_t;
```
“Data” is vague. Should this type be called `psa_pake_step_t`? I'm not a great fan of “step”, but it's what we settled on for key derivation, and it's a very similar concept here.

gilles-peskine-arm commented on include/psa/crypto_values.h:
```diff
@@ -2011,4 +2097,154 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ */
+#define PSA_PAKE_SIDE_SERVER                ((psa_pake_side_t)0x12)
+
+/** The PAKE uses elliptic curves.
```
This description is missing an indication of the kind of entity that's being described. Maybe something like this:
```suggestion
/** The PAKE primitive type indicating the use of elliptic curves.
```

gilles-peskine-arm commented on include/psa/crypto_values.h:
```diff
@@ -2011,4 +2097,154 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+
+/** The PAKE uses elliptic curves.
+ *
+ * The corresponding family type is ::psa_ecc_family_t. In determining a
+ * specific curve in the family the cipher suite (see
+ * ::psa_pake_cipher_suite_t) bits are interpreted in the exact same way
+ * as key bits are.
```
I know what you mean, but I can't parse this sentence.

gilles-peskine-arm commented on include/psa/crypto_values.h:
```diff
@@ -2011,4 +2097,154 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ * Some PAKE algorithms need to exchange several key shares. If that is the
+ * case, this value marks the first key share sent and the first key share
+ * received. For values sent or received afterwards, use
+ * #PSA_PAKE_DATA_KEY_SHARE_2 and #PSA_PAKE_DATA_KEY_SHARE_3.
```
Do we need `PSA_PAKE_DATA_KEY_SHARE_2` and `PSA_PAKE_DATA_KEY_SHARE_3`, or can we say that the first share is the first call to `psa_pake_input with `PSA_PAKE_DATA_KEY_SHARE`, the second share is the second call, etc.?

athoelke replied:
I have to agree that this would be neat.

Presumably, PAKE constructions that have multiple key shares will always consume and produce the key shares in numerical order? - so using `PSA_PAKE_DATA_SHARE_X` would demand step-sequence validation, and provides no functional utility over @gilles-peskine-arm's suggestion of a single data type/step.

gilles-peskine-arm reviewed (changes requested):
This is converging but there are still a few hurdles, especially with respect to passing the many parameters required to set up and perform a PAKE.

athoelke commented on include/psa/crypto.h:
```diff
@@ -3759,6 +3760,483 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *
+ * \param[in] cipher_suite      The cipher suite structure to query.
+ *
+ * \return The hash algorithm stored in the cipher suite structure.
```
Could use `PSA_ALG_NONE` here instead of zero? - currently being included via the PBKDF PR.

mpg replied:
Actually Gilles was not comfortable introducing it now, so I've just removed that.

gilles-peskine-arm commented:
Now that the PBKDF2 is merged, there are some merge conflicts. Please rebase or merge at your convenience. The definition of `PSA_KEY_TYPE_PASSWORD` is available now.

athoelke commented on include/psa/crypto_values.h:
```diff
@@ -1939,6 +1952,84 @@
+ * J-PAKE doesn't differentiate between sides and has a distinct identifier
+ * associated with each entity. This means that both psa_pake_set_user() and
+ * psa_pake_set_peer() needs to be called before commencing the operation.
+ * Another conseque is that psa_pake_set_side() is optional and is ignored.
```
```suggestion
 * Another consequence is that psa_pake_set_side() is optional and is ignored.
```

athoelke commented on include/psa/crypto.h:
```diff
@@ -4009,6 +3974,545 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+                           psa_algorithm_t algorithm
+                           );
+
+/** Retrieve the primitive type from a PAKE cipher suite.
```
Is this still the "primitive type" or should the summary now use just "primitive"?

athoelke commented on include/psa/crypto.h:
```diff
@@ -4009,6 +3974,545 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+                           const psa_pake_cipher_suite_t* cipher_suite
+                           );
+
+/** Declare the primitive type for a PAKE cipher suite.
```
Same comment as for `psa_pake_cd_get_primitive()`

athoelke reviewed (commented):
Overall, looking good now. I have some questions on the new MHF password input function.

yanesca commented on include/psa/crypto.h:
```diff
@@ -4009,6 +3974,545 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *                              not yet in use.
+ * \param key_derivation        An ongoing key derivation operation set up from
+ *                              the password and in a state suitable for
+ *                              calling psa_key_derivation_output_bytes().
```
Good point. I just checked and the derivation doesn't know. There is no optimal or recommended size, PAKEs don't mandate it either (with the exception of OPAQUE, but that uses it internally and won't be using this interface). It makes most sense if this is application controlled.

athoelke replied:
For consistency, this function needs to include the same paragraph that is present on other key derivation operation extraction functions (output_xxx() and verify_xxx()) regarding the consumption of `input_length` bytes of the output of the operation.

yanesca commented on include/psa/crypto.h:
```diff
@@ -4009,6 +3974,545 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *         results in this error code.
+ */
+psa_status_t psa_pake_set_password_mhf(psa_pake_operation_t *operation,
+                                       psa_pake_operation_t *key_derivation);
```
Since both of these are operation objects technically both are [in,out] parameters. `operation` takes an input from `key_derivation`, but its state changes as well. The same thing happens for `psa_pake_get_implicit_key()`, just the way around.

From the perspective of thinking in objects both of these functions are better thought of as related to the PAKE operation, because they need direct/low level access to the PAKE structure, whereas (in theory) they can use the public API of the key derivation operation without any need for low level access to the key derivation structure.

I put the PAKE operation in both parameter lists first, because that is the object that these functions belong to. And by "belong to" I mean that in other languages this wouldn't need to be passed and would be referred to as `self` or `this`. In cases like this I think it makes sense to put the "self/this" structure first in the parameter list.

yanesca commented:
@athoelke , @gilles-peskine-arm Thank you both very much for the reviews!

I believe I addressed all of your feedback and the PR is ready for review again.

gilles-peskine-arm reviewed (changes requested):
I reviewed the whole pull request with a main focus on the application developer's perspective. I think the API design is ok except maybe for a few names. There are several places where the specification could use some clarification.

yanesca commented on include/psa/crypto.h:
```diff
@@ -4009,6 +3974,565 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
 /**@}*/
 
+
+/** \defgroup Password-authenticated key agreement
```
I guess then I should move everything PAKE related into `crypto_extra.h`.

gilles-peskine-arm replied:
Yes, I think that's easy to do and clarifies our intentions. That plus adding a note at the top of the PAKE section, because `crypto_extra.h` also contains things that are Mbed TLS extensions but for which we do promise stability.

yanesca commented on include/psa/crypto.h:
```diff
@@ -4297,63 +4184,218 @@ static psa_pake_operation_t psa_pake_operation_init(void);
+ *                              must have been set up by psa_pake_setup() and
+ *                              not yet in use.
  * \param password              Identifier of the key holding the password or a
  *                              value derived from the password (eg. by a
```
It is entirely the application's choice: algorithms treat the password as a byte string and usually hash it before use.

yanesca commented on include/psa/crypto.h:
```diff
@@ -4261,23 +4138,33 @@ static psa_pake_operation_t psa_pake_operation_init(void);
- *    identities and additional session information.
+ * -# Call psa_pake_setup() to specify cipher suite.
+ * -# Call \c psa_pake_set_xxx() functions on the operation to complete the
+ *    setup. The exact sequence of \c psa_pake_set_xxx() functions that needs
```
In some cases the implementation might be able to do some optimisations if the parameters arrive in a given order, but that might be dependent on the algorithm.

In the case of JPAKE the order doesn't matter.

yanesca commented on include/psa/crypto.h:
```diff
@@ -4297,63 +4184,218 @@ static psa_pake_operation_t psa_pake_operation_init(void);
+ * \retval #PSA_SUCCESS
+ *         Success.
+ * \retval #PSA_ERROR_BAD_STATE
+ *         The operation state is not valid (it must have been set up.)
```
> This description is for the PAKE operation, but there's also a KDF operation which must be in the right state as well.

In this case the call to `psa_key_derivation_output_bytes()` returns `PSA_ERROR_BAD_STATE` (which is something other than `PSA_ERROR_INSUFFICIENT_DATA`) and thus we return `PSA_ERROR_INVALID_ARGUMENT`. From the perspective of this function the KDF operation is an argument and the PAKE operation is the ongoing operation.

> For the PAKE, it's also an invalid state if psa_pake_output or psa_pake_input has already been called.

Good point, will fix.

athoelke replied:
You have already stated that this function interacts with (and even modifies) two operation objects, and either could be in the wrong state. Cloaking a `BAD_STATE` error from the kdf does not help the application developer - as this is a key indicator that they have a bug in their code, rather than some transient issue.

Just about any error returned by the function leaves the two operation objects in an uncertain state, and the best advice will be for the application to abort both of them - it is hard to work out how any other approach would be reliable in this circumstance.

I expect that all of this commentary also applies to `psa_pake_get_implicit_key()`.

yanesca replied:
> Just about any error returned by the function leaves the two operation objects in an uncertain state, and the best advice will be for the application to abort both of them - it is hard to work out how any other approach would be reliable in this circumstance.

Sounds good, will fix.

yanesca commented on include/psa/crypto.h:
```diff
@@ -4297,63 +4184,218 @@ static psa_pake_operation_t psa_pake_operation_init(void);
+psa_status_t psa_pake_set_password_key(psa_pake_operation_t *operation,
+                                       mbedtls_svc_key_id_t password);
+
+/** Set the password for a password-authenticated key exchange via a memory hard
+ * function.
```
> We're using “key stretching” as the generic term for such functions.

Thanks, will do.

> I think a better name for this function would be `psa_key_derivation_output_for_pake`, with the key derivation operation parameter (the source) coming before the pake operation parameter (the destination).

Both of these are operation objects technically [in,out] parameters. `operation` takes an input from `key_derivation`, but  the state of `key_derivation` changes as well.

I think this function is better thought of as related to the PAKE operation, because it needs direct/low level access to the PAKE structure, whereas (in theory) it can use the public API of the key derivation operation without any need for low level access to the key derivation structure.

athoelke replied:
After discussion during the Crypto call today, I am happy to go with Gilles suggestion to organise multi-operation APIs with input operations first, followed by output operations. This is considering the logical usage of the operation, not the `const`ness of the operation.

This ordering is systematically used for the non-operation parameters in all PSA Crypto functions, so it is consistent to continue that for APIs such as this.

I think this is a stronger argument for making this a `psa_key_derivation_xxx()` function than whether or not the implementation can use an existing published API to achieve the same effect.

`psa_key_derivation_output_for_pake()` is kind of clunky as a name, and isn't specific as to which part of the PAKE inputs are being provided, and PAKE takes a lot of them. Would `psa_key_derivation_output_pake_password()` be better?

gilles-peskine-arm replied:
I don't like `psa_key_derivation_output_pake_password` because it makes it look like it's outputting the password. But the password is actually an input to the key derivation operation. What's being input into the PAKE is a value derived from the password.

athoelke replied:
I finally took the time tot read the J-PAKE spec material, and the publically available THREAD material (which happens to be the mbedtls implementation in OpenThread :wink:).

The input to the J-PAKE algorithm is a [shared] low entropy secret _s_, which is described as being a password or hash of a password. J-PAKE does not require this to be a memory- or computationally- hard process - and that would potentially undermine THREAD's use of J-PAKE in IoT devices. A cryptographic hash of the password is suggested for J-PAKE in order to ensure that _s_ can be converted to an integer for use in the key establishment process without placing restrictions on the password itself to meet the J-PAKE requirements for _s_.

This leads to a couple of observations/questions:

1. The secret/password input to J-PAKE is unlikely to be a password-hash (PBKDF2-like), so this API is not that relevant for J-PAKE?
2. Should we use 'secret' in the interface names and description for this input to a PAKE, rather than 'password'? - we can describe the use of passwords, cryptographic hashes of passwords, or full password-hash in the full description, or in the individual algorithm descriptions?

gilles-peskine-arm replied:
If this function is not useful for PAKE, and given that we aren't sure about its design, I'm for leaving it out.

> Should we use 'secret' in the interface names and description for this input to a PAKE, rather than 'password'?

We're already calling the output of a PAKE a “shared secret”. The password is a shared secret (for PAKE where the verifier knows a password-equivalent secret), but calling it “shared secret” or even just “secret” would be confusing.

yanesca replied:
> If this function is not useful for PAKE, and given that we aren't sure about its design, I'm for leaving it out.

I was on the same opinion and it is already removed in the current version.

yanesca replied:
(It was probably a typo, but just to clarify: it is useful for PAKE, just not for J-PAKE.)

yanesca replied:
> We're already calling the output of a PAKE a “shared secret”. The password is a shared secret (for PAKE where the verifier knows a password-equivalent secret), but calling it “shared secret” or even just “secret” would be confusing.

Good point, I hadn't thought about that. Reverted the change.

athoelke commented on include/psa/crypto.h:
```diff
@@ -4009,6 +3974,565 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ * \retval #PSA_ERROR_STORAGE_FAILURE
+ * \retval #PSA_ERROR_NOT_PERMITTED
+ * \retval #PSA_ERROR_INVALID_ARGUMENT
+ *         The call to psa_key_derivation_output_bytes() returned something
```
Exposing just one possible implementation option. If this happens inside the cryptoprocessor, there is no "call to `psa_key_derivation_output_bytes()`" (note function name not monospace or linked).

But see other comments that errors coming from the KDF should probably just be propogated out to the caller.

yanesca replied:
Good point, haven't thought about that. I think that applying the approach to error handling you suggested above would fix this as well.

athoelke commented on include/psa/crypto.h:
```diff
@@ -4009,6 +3974,565 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ * state and must be aborted by calling psa_pake_abort().
+ *
+ * \param[in,out] operation    Active PAKE operation.
+ * \param type                 The type of the data that is requested.
```
`type` has the type `psa_pake_step_t` - so the name is inconsistent as well as confusing. Could we just call it `step`?

yanesca replied:
I agree, I have already changed the name to `step` in the fix.

athoelke commented on include/psa/crypto.h:
```diff
@@ -4009,6 +3974,565 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *
+ * When this function returns successfully, the operation becomes inactive.
+ * If this function returns an error status, the operation enters an error
+ * state and must be aborted by calling psa_pake_abort().
```
Which operation enters an error state? - as we do not provide enough information in the return code, and we do provide an operation status query function, the caller has to assume that both operation objects will need to be aborted.

yanesca commented on include/psa/crypto.h:
```diff
@@ -4297,63 +4184,218 @@ static psa_pake_operation_t psa_pake_operation_init(void);
+ * \retval #PSA_SUCCESS
+ *         Success.
+ * \retval #PSA_ERROR_BAD_STATE
+ *         The operation state is not valid (it must have been set up.)
```
`set_peer` and `set_user` should be fairly independent, although in the case of some algorithms calling them in a particular order could allow the implementation to do some minor optimisations.

yanesca commented on include/psa/crypto.h:
```diff
@@ -4261,23 +4138,33 @@ static psa_pake_operation_t psa_pake_operation_init(void);
- *    times to exchange any further data that is needed to derive the shared
- *    secret.
+ * -# Make a sequence of function calls to execute the password-authenticated
+ *    key exchange as described below.
```
This was a remnant of a previous description and should have been deleted already.

athoelke commented on include/psa/crypto.h:
```diff
@@ -4009,6 +3974,565 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *                              memory-hard function).  It must remain valid
+ *                              until the operation terminates. It must be of
+ *                              type #PSA_KEY_TYPE_PASSWORD or
+ *                              #PSA_KEY_TYPE_DERIVE. It has to allow the usage
```
Should this accept a key of type `PSA_KEY_TYPE_PASSWORD_HASH` as well, or instead of `PSA_KEY_TYPE_DERIVE`?

yanesca commented on include/psa/crypto_values.h:
```diff
@@ -1939,6 +1952,84 @@
         (((alg) & 0x000000ff) == 0 ? ((psa_algorithm_t)0) : 0x02000000 | ((alg) & 0x000000ff))
 
+/** The Password-authenticated key exchange by juggling (J-PAKE) algorithm.
+ *
```
I reorganised the description as per your later comments, I hope it satisfies this one as well.

yanesca commented:
The CI is only failing on the head because of the removed generated file, which is expected.

yanesca commented:
@gilles-peskine-arm , @athoelke , @stevew817 Thank you very much for your feedback! I believe I have addressed all of your comments, could you please review the PR again?

gilles-peskine-arm commented on include/psa/crypto_struct.h:
```diff
@@ -93,10 +93,10 @@ struct psa_hash_operation_s
 
 #define PSA_HASH_OPERATION_INIT {0, {0}}
-static inline struct psa_hash_operation_s psa_hash_operation_init( void )
+static inline struct psa_hash_operation_s psa_hash_operation_init(void)
```
Not a blocker, but the commit “Fix paranthesis spacing in crypto_struct.h” (sic) shouldn't be here. `crypto_struct.h` is implementation-specific so it was intended to have Mbed TLS whitespace style.

yanesca replied:
I did that change to address one of your comments. The spacing was mixed and I didn't see  any indication of the intended whitespace style and that is why I aligned it with the style I saw in `crypto.h`.

I will change it.

gilles-peskine-arm commented on include/psa/crypto_extra.h:
```diff
@@ -813,6 +813,987 @@ psa_status_t mbedtls_psa_platform_get_builtin_key(
+ *                      compatible with algorithm \p alg.
+ * \param input_step    A value of type ::psa_pake_step_t that is valid for the
+ *                      algorithm \p alg.
+ * \return              A sufficient output buffer size for the specified
+ *                      output, cipher suite and algorithm. If the cipher suite,
+ *                      the output type or PAKE algorithm is not recognized, or
+ *                      the parameters are incompatible, return 0.
```
“output” copypasta is still present here.

gilles-peskine-arm commented on include/psa/crypto_extra.h:
```diff
@@ -813,6 +813,987 @@ psa_status_t mbedtls_psa_platform_get_builtin_key(
+
+/** A sufficient input buffer size for psa_pake_input().
+ *
+ * If the size of the input is larger than this, it is guaranteed
+ * that psa_pake_input() will fail with #PSA_ERROR_INVALID_ARGUMENT.
```
This description is correct, but it's a weird way to put it, and I think out of context most readers would wonder why it's useful to have a guarantee that an API function will fail. Suggested explanation:

> The value returned by this macro is guaranteed to be large enough for any valid input to psa_pake_input() in an operation with the specified parameters.

gilles-peskine-arm commented on include/psa/crypto_extra.h:
```diff
@@ -813,6 +813,987 @@ psa_status_t mbedtls_psa_platform_get_builtin_key(
+ * keys on the group determined by the primitive (::psa_pake_primitive_t) would
+ * be.
+ *
+ * Some public key algorithms mask the private keys and this might be reflected
+ * in the export format. Even if this is the case the masking is omitted at
+ * this step.
```
I don't understand what “the masking is omitted” means. I presume it's referring to Montgomery keys. The previous paragraph states that the format is exactly the export format, which only allows the canonical form. Do you mean that the ZK proofs can have non-canonical representations of Montgomery keys? If so the previous paragraph is wrong.

Can we require canonical representations at this point, or would this make it impossible to have interoperable transcripts?

gilles-peskine-arm replied:
Ok, so the specification needs to describe the non-canonical formats that are accepted in input and the non-canonical formats that may be produced in output. When multiple representations are possible, does the choice of representation that our implementation makes make a difference to the peer?

yanesca replied:
It might or might not make a difference. Some implementations might not be able to handle non-canonical formats (eg I would expect this from some hardware accelerators) and they would need to convert.

On the other hand if we prescribe canonical format, everyone needs  to convert.

I chose this way of describing it because I wanted to make it more universal. Of course we can simply say that it is parsed as in SEC 1 with the exception of Montgomery curves, where the byte order is reversed. (This wouldn't change the meaning.)

gilles-peskine-arm commented on include/psa/crypto_extra.h:
```diff
@@ -813,6 +813,987 @@ psa_status_t mbedtls_psa_platform_get_builtin_key(
+ * guarantee that the peer is the party he claims to be and was able to do so.
+ *
+ * That is, the authentication is only implicit (the peer is not authenticated
+ * at this point, and no action should be taken that assume that they are - like
+ * for example accessing restricted files).
```
Nitpicks:
* `-` is a hyphen, you mean an em-dash `—`.
* “like for example” is redundant: both “like” and “for example” indicate that an example follows.

Suggestion to simplify this sentence:

> That is, the authentication is only implicit. Since the peer is not authenticated yet, no action should be taken yet that assumes that the peer is who it claims to be. For example, do not access restricted files on the peer's behalf until an explicit authentication has succeeded.

gilles-peskine-arm commented on include/psa/crypto_extra.h:
```diff
@@ -813,6 +813,987 @@ psa_status_t mbedtls_psa_platform_get_builtin_key(
+ *
+ * At this point there is a cryptographic guarantee that only the authenticated
+ * party who used the same password is able to compute the key. But there is no
+ * guarantee that the peer is the party he claims to be and was able to do so.
```
Please avoid gendered language (“he”) if possible. Neutral “he” used to be acceptable in English, but in the 21st century this has become frowned upon. “It” is ok for a peer in a protocol.

gilles-peskine-arm commented on include/psa/crypto_extra.h:
```diff
@@ -813,6 +813,987 @@ psa_status_t mbedtls_psa_platform_get_builtin_key(
+ *      psa_pake_set_peer(operation, ...);
+ *      psa_pake_set_password_key(operation, ...);
+ *
+ * The password is read as a byte array and must be non-empty. This can be the
```
Why would an empty password not be supported? I don't have a concrete example, but it seems plausible that a protocol with an optional password would have a mandatory PAKE to keep the software design simple, and would use an empty password if the security policy said that a password is not used.

A precedent for allowing an empty secret input is in key derivation, with a concrete use case of using the KDF algorithm to derive both a key and an IV, with an empty “secret” input to the KDF when deriving the IV. For this case, `psa_key_derivation_input_key` can't be used because keys never have zero length, but `psa_key_derivation_input_bytes` does accept an empty input.

athoelke replied:
This might be because J-PAKE specifically requires the secret  _s_ &ne; 0? For J-PAKE, the secret is an integer - and a common conversion from a byte-string to an integer results in the value zero for the empty string.

gilles-peskine-arm replied:
Good point, makes sense, so you can't to this with J-PAKE. But I think this should be a constraint for `PSA_ALG_JPAKE` rather than a mandatory constraint for all algorithms.

yanesca replied:
J-PAKE does not prescribe how we make an integer of the password though. If the application or higher level protocol chooses to hash the password and convert it to an integer, then empty passwords become feasible.

(If the secret integer `s` is 0 mod the group order, then the key material will always be the generator element.)

yanesca replied:
> But I think this should be a constraint for `PSA_ALG_JPAKE` rather than a mandatory constraint for all algorithms.

I agree. As far as I can tell, this part is specific to `PSA_ALG_JPAKE` and the description of `psa_pake_set_password_key()` doesn't mention such restrictions.

athoelke replied:
Given that J-PAKE expects the secret to be a big integer, and the PAKE API only accepts a key (which is typically considered to be an octet string) as input - this specification needs to document how the key material is converted to the integer _s_.

That conversion is not specified by J-PAKE, it only hints at some application options. As this is a two party protocol, and we shouldn't assume PSA Crypto at the other end, we need to enable any reasonable approach. That suggests that we should use the most obvious "octet-string is the big*-endian representation of an integer" (* or little?), which allows application flexibility (to hash or not to hash).

yanesca replied:
It is documented in the next paragraph:
https://github.com/ARMmbed/mbedtls/pull/4317/files#diff-8fe19c51d861f48010b3609673ee1dd94bb3eef20af920af31b36fc0dcb124c8R863-R868

yanesca replied:
(This allows for an application to do hashing beforehand if they wish or any other transformation that is required.)

gilles-peskine-arm reviewed (changes requested):
We're almost there but there are a couple of questions left about the details of the semantics.

yanesca commented on include/psa/crypto_extra.h:
```diff
@@ -813,6 +813,987 @@ psa_status_t mbedtls_psa_platform_get_builtin_key(
+ * - The group can be either an elliptic curve or defined over a finite field.
+ * - Schnorr NIZK proof as defined by RFC 8235 and using the same group as the
+ *   J-PAKE algorithm.
+ * - A secure cryptographic hash function.
```
I was using this expression, because this is exactly how it stands in the standard. I wasn't thinking about explicitly forbidding SHA-1. But now that I think of it, I can't see any obvious reasons why SHA-1 shouldn't be used (intuitively we need a PRF property here), but the security proof assumes it to be a random oracle, which of course impossible to satisfy, but having collision resistance as well is one step closer.

Long story short: I didn't mean to make a stance on SHA-1 with this statement, but it is not a catastrophe if that is how it is interpreted.

yanesca commented:
I have rebased to resolve conflicts before starting the rework:
- "PSA PAKE: add to PSA constant name test": Changes moved to new file, new change needed (0b055c763 -> 8603fb09f)
- "PSA PAKE: generate PSA constants" and its revert became empty and are omitted (2a5fe055b and d1a682080)
- "Fix parenthesis spacing in crypto_struct.h" fixed typo in commit title (d23be468e -> a9b5d1ba5)

I have saved the old branch here:
https://github.com/yanesca/mbedtls/tree/psa-jpake-interface-pre-rebase-3

gilles-peskine-arm reviewed (commented):
I have verified that 8b73430 to e4b58db is a good rebase on top of an updated version of the target branch.

yanesca commented:
@gilles-peskine-arm I believe I addressed all of your comments and the PR is ready for review again.

stevew817 reviewed (dismissed):
I have re-reviewed the entire PR, and don't have anything to add other than that CI seems to be slightly unhappy.

yanesca commented:
@stevew817 Thanks for the review and for the reminder! (I didn't notice that the CI was failing.)

gilles-peskine-arm commented on include/psa/crypto_extra.h:
```diff
@@ -813,6 +813,1005 @@ psa_status_t mbedtls_psa_platform_get_builtin_key(
+ * This is the proof in the Schnorr Non-Interactive Zero-Knowledge Proof (the
+ * value denoted by the letter 'r' in RFC 8235).
+ *
+ * Both input and output at this step is an integer less than the group
```
Minor: “both” needs a plural verb.
```suggestion
 * Both input and output at this step are integers less than the group
```
or
```suggestion
 * Each of input and output at this step is an integer less than the group
```
or
```suggestion
 * Both for input and output, the value is an integer less than the group
```

gilles-peskine-arm replied:
Missing word? “integer less than the group ??”

gilles-peskine-arm commented on include/psa/crypto_extra.h:
```diff
@@ -813,6 +813,1005 @@ psa_status_t mbedtls_psa_platform_get_builtin_key(
+ * Both input and output at this step is an integer less than the group
+ * selected in the cipher suite. The format depends on the group as well:
+ *
+ * - For Montgomery curves, the encoding is little endian (see decoding of
+ *   scalars in Section 5 of RFC 7748).
+ * - For everything else the encoding is big endian (see Section 2.3.8 of
+ *   _SEC 1: Elliptic Curve Cryptography_ at https://www.secg.org/sec1-v2.pdf).
```
(Continuing https://github.com/ARMmbed/mbedtls/pull/4317#discussion_r643020180) The format is still not fully specified. What about the size? Are there leading zeros?

yanesca commented:
Rebased to resolve https://github.com/ARMmbed/mbedtls/pull/4317#discussion_r646730668.

No intended code changes. Branch before rebase:
https://github.com/yanesca/mbedtls/tree/psa-jpake-interface-pre-rebase-4

yanesca commented on include/psa/crypto_extra.h:
```diff
@@ -813,6 +813,1005 @@ psa_status_t mbedtls_psa_platform_get_builtin_key(
+ * the official PSA Crypto API yet.
+ *
+ * \note The content of this section is not part of the stable API and ABI
+ *       of Mbed Crypto and may change arbitrarily from version to version.
```
We merged the repository back, but it is technically still Mbed Crypto. I don't have a strong opinion of stopping referring to Mbed Crypto in the documentation and the code, but that decision is out of scope for this PR  and if it is made I would prefer to do it throughout the library.

yanesca commented:
@gilles-peskine-arm I believe I have addressed all of your comments. The PR is ready for review again.

gilles-peskine-arm reviewed (changes requested):
LGTM apart from one issue.

yanesca commented on include/psa/crypto_extra.h:
```diff
@@ -1130,14 +1130,17 @@ typedef uint32_t psa_pake_primitive_t;
  * - For everything else the encoding is big endian (see Section 2.3.8 of
  *   _SEC 1: Elliptic Curve Cryptography_ at https://www.secg.org/sec1-v2.pdf).
  *
+ * In both cases leading zeroes are allowed as long as the length in bytes does
+ * not exceed the byte length of the group order.
```
That is true for secret values. These are public. They are very much like signatures: verifying a signature proves that the holder of the private key signed the message. Verifying the ZK proof proves that the holder has the key and nothing else.

gilles-peskine-arm replied:
Ah, yes, good point, I was looking at this sentence in isolation and hadn't realized it only applied to public values.

Is there any advantage in allowing output to have leading zeros truncated though?

yanesca replied:
> Is there any advantage in allowing output to have leading zeros truncated though?

Reduced cognitive load both on the user's and library implementor's part and maybe a marginal code size advantage since the implementation needs to do one less check.

gilles-peskine-arm replied:
I think having a predictable data size is easier for both applications and implementations. But it's minor and we can change it later.

(I'm reminded of a security presentation: “The protocol included one variable-length field. So we tried a stupidly long input. The parser crashed.”)

gilles-peskine-arm reviewed (approved):
All right, let's go with it!

yanesca commented:
@gilles-peskine-arm Thank you for approving!

@stevew817 Thank you for your review! Could you please have a look at the changes made since your last approval and re-approve?
